### ISSUE TITLE
FM-1 — Production two-step checkpoint converter (legacy pickle → neutral artifact → v1) and end-user migration tool

### ISSUE THREAD

**Issue by aacostadiaz** [ACEsuit/mace#1586]
**Depends on:** FM-00 (#1565) · **Blocks:** FM-2 (#1587), JAX-1 (#1589)

**Prerequisite:** the Phase 0 characterization suite is committed — goldens, fixtures and the single-source tolerance table under `tests/golden/`, characterization tests under `tests/unit/`. References of the form `P0-N` name that work.

**Context:** FM-00 (#1565) proved the conversion concept on one committed tiny anchor: running **in-process** against the in-tree frozen `mace/`, it extracts the anchor to the neutral format, maps it onto the v1 architecture, and reproduces the Phase 0 E/F/stress goldens at fp64. FM-1 (#1586) generalizes that keystone into the **production converter** — the tool FM-2 (#1587) runs over every published foundation-model artifact, and the tool end users run on their own v0.3.x checkpoints. By decision, v0.3.x checkpoints are **not runtime-loadable** in v1; "loadable" means *one-shot convertible* with this tool, and the eventual migration guide (REL-1 (#1603)) points here — for LAMMPS users the instruction becomes "convert the checkpoint, then export". Legacy checkpoints are **full Python pickles**: loading one executes legacy class definitions (which is why `mace/__init__.py` sets `TORCH_FORCE_NO_WEIGHTS_ONLY_LOAD=1` — import e3nn before `mace` and even `e3nn`'s own `constants.pt` fails to unpickle), so the extraction step needs a live legacy `mace` importable — in-tree during development, or an isolated pinned environment for end users who only have v1 installed. The v1 side must never unpickle anything.

**Interface & constraints:**

1. **Two steps, one interface file.** Step 1 (**extract**, legacy side): unpickle the checkpoint with legacy code, map every weight to the canonical layout, write the neutral artifact. Step 2 (**import**, v1 side): read the neutral artifact, build the v1 model. The neutral artifact is the *only* thing that crosses the boundary; step 2 imports only `packages/` code, never `mace/`.

2. **Two operating modes for step 1:**

| Mode | Legacy source | Isolation | Used by |
|---|---|---|---|
| development | in-tree frozen `mace/`, imported in-process | none (no venv) — the FM-00 (#1565) path | FM-2 (#1587) batch conversion, parity tests, CI |
| end-user packaging | pinned venv: `mace-torch==0.3.17` floor for published FM pickles (**verify the published artifacts still deserialize under it** before REL-1 (#1603) ships the guide) | subprocess in the venv; only the neutral file crosses | users migrating their own v0.3.x checkpoints |

   The venv mode exists because an end user's v1 install has no legacy `mace` to unpickle against; the pinned version keeps the unpickling environment reproducible and keeps pickle execution out of the v1 process.

3. **Neutral artifact = safetensors + JSON sidecar, no pickle.** The safetensors file holds all canonical tensors, flat-keyed; the JSON sidecar holds the model config, one `CanonicalSpec` per op (`op_kind`, `schema_version`, `clebsch_gordan_basis`, `parametrization` — `"nested"`/`"fused"` — optional `proj` change-of-parametrization matrix, serialized descriptor fields), precision, and graph structure. Canonical layout is `mul_ir` with first-class bias, flat `[Z, A, mul]` for the symmetric contraction (the CORE-4 (#1558)-pinned path order and per-path normalization *is* the on-disk weight format). Never a backend `state_dict`.

4. **Supported source families, and refusal by name for everything else.** The converter handles the energy family — `MACE` (`mace/modules/models.py:47`) and `ScaleShiftMACE` (`:444`) — plus `AtomicDielectricMACE` (`:842`), which FM-2 (#1587) needs for MACE-MDP. Every other class in today's model registry is **refused by name**, with the owning ticket in the message: `AtomicDipolesMACE` (`:626`) and `EnergyDipolesMACE` (`:1199`) → the observables-as-heads path (ARCH-3 (#1562)); `PolarMACE` (`mace/modules/extensions.py:663`) → FM-2 (#1587), which adds the polar source path on top of ELEC-2 (#1592)'s registered model; `MACELES` (`:142`) → ELEC-4 (#1635); `MagneticMACE` (`:1428`), `MagneticScaleShiftMACE` (`:1706`) and `MagneticSCFMACE` (`:1968`) → MAG-1 (#1584). Refusal is by class identity, not by duck-typing the weight names: a magnetic checkpoint's energy blocks look convertible right up to the point where its magmom basis is silently dropped.

5. **Full metadata, measured against what legacy actually reconstructs.** `mace/tools/scripts_utils.py:230-388` (`extract_config_mace_model`) is the authoritative list of what it takes to rebuild a legacy model with no legacy code: it yields **33 keys** for each of MACE-MP-0 small, MACE-MPA-0 medium and MACE-OFF23 medium, and 32 for the MACE-MDP `AtomicDielectricMACE`. The sidecar carries all of them — including the ones a shorter list would drop and nothing would notice until parity fails: `avg_num_neighbors`, `apply_cutoff`, `radial_MLP`, `distance_transform`, `pair_repulsion`, `use_so3`, `use_agnostic_product`, `use_last_readout_only`, `use_embedding_readout`, `use_edge_irreps_first`, `edge_irreps`, `embedding_specs`, `correlation`, `gate` — plus heads, atomic numbers, E0s, r_max, scale/shift, dtype/precision, and provenance: source-file hash, source `mace-torch` version, converter version. Key-count equality against `extract_config_mace_model` on the source model is an acceptance item, not a guideline.
   **Headless artifacts are the common case, not an edge case.** MACE-MP-0 small and both MACE-OFF23 tiers carry no `heads` attribute at all; legacy substitutes `["default"]` in the config-extraction path (`scripts_utils.py:263`) and `["Default"]` in others (`mace/calculators/mace_torchsim.py:145`, `mace/modules/extensions.py:713`). The converter follows the config-extraction path and records `["default"]`, plus a flag saying the source was headless — so a later reader can tell a genuinely single-head model from one that predates heads. Multi-head artifacts convert with **all heads intact**; single-head export is a separate `mace model select-head` step (CLI-1 (#1579)), not a converter behaviour.

6. **Basis rules (settled, converter must enforce).** Develop already has a first-class reduced Clebsch-Gordan basis, so this is a conversion between two live formats, not the introduction of one:
   - **What develop implements.** `--use_reduced_cg` (`mace/tools/arg_parser.py:239-244`, default **`False`**) → `MACE.use_reduced_cg` (`mace/modules/models.py:67,103`) → `SymmetricContraction(use_reduced_cg=…)` → `U_matrix_real(use_cueq_cg=…)` (`mace/tools/cg.py:106-140`), and it is serialized by `scripts_utils.py:293-295`. The reduced basis is cuequivariance's `compute_U_cueq`, reachable only when `cuequivariance` imports (`CUET_AVAILABLE`, `cg.py:17-21`); the process-wide default comes from `MACE_USE_CUEQ_CG` (`cg.py:23-28`). So today's *default* trained model is full-basis, and reduced is an opt-in that needs an optional dependency.
   - **Read `clebsch_gordan_basis` from the artifact, never assume it.** Every published foundation model is a full-basis, nested-parametrization artifact: none carries a `use_reduced_cg` attribute, so `scripts_utils.py:293-295` records `False` for all of them, and their contraction weights hold the full path counts (below).
   - **The pinned path counts.** For the published coupling `1x0e+1x1o+1x2e+1x3o` at `correlation=3`, per contraction and per ν=1,2,3: a `0e` output is `1+4+23 = 28` full → `1+4+8 = 13` reduced; a `1o` output is `1+6+51 = 58` full → `1+3+12 = 16` reduced. MACE-MP-0 small is two product blocks with one `0e` contraction each (28 → 13); MACE-OFF23 medium and MACE-MPA-0 medium are block 0 = `0e`+`1o` (86 → 29) and block 1 = `0e` (28 → 13). These per-`(irrep_out, ν)` dimensions are the compatibility target: v1's own reduced basis must have exactly them.
   - **Emit reduced-basis artifacts.** The projection is exact at fp64 and removes real weight: −341,760 of 3,847,696 params on MP-0 small (**8.9 %**), −820,224 of 9,063,204 on MPA-0 medium (**9.1 %**), −92,160 of 1,428,368 on OFF23 medium (**6.5 %**), with unchanged predictions.
   - **Build the projection from the symmetrized map — this is the trap.** The extra full-basis directions are gauge *only on the symmetric diagonal*, because the contraction feeds the same node feature to every leg. Fitting `U.reshape(-1, paths)` treats the legs as independent, finds full rank (23 at ν=3 for a `0e` output, not 8) and leaves a residual of order 0.4 — a projection that looks broken rather than a wrong formulation. Fitting the map `w ↦ Σ_p U[…,p] · x⊗…⊗x` over sampled feature vectors `x` gives exactly the reduced rank and a relative max residual ~2e-15 at fp64. The projection is one small matrix per irreps configuration, reused across elements and channels (weights are `[element, path, channel]`; the contraction acts per `(element, channel)`).
   - The projection is **fitted at conversion time against v1's own reduced basis**, so v1's basis need not share cuequivariance's gauge — only its dimensions. It is **function-preserving, not bit-exact in weight space**: verification gates on E/F/stress outputs at fp64 tolerance, **never** on weight equality.
   - **Refuse `reduced → full`.** It is under-determined (min-norm representative only); full-basis *export* for external reimplementations is DEP-3 (#1595)'s job, not this tool's.
   - **Recorded deviation:** develop's `use_reduced_cg=True` on a host without `cuequivariance` falls through to the full basis with no diagnostic, so a model's basis silently depends on what is installed. v1 reproduces the *numerics* of both bases and does **not** reproduce the silent fallback — a requested basis that cannot be built is a hard error.

7. **Golden verification hook.** The converter takes an optional reference snapshot in the framework-agnostic golden-harness JSON format (from P0-1's `tests/golden/harness.py`; numpy/ase/json only, importable by both stacks) and, after import, runs the converted v1 model on the snapshot's fixtures and compares E/F/stress at fp64 tolerance — failing the conversion loudly on breach. In development mode it can additionally self-generate the reference from the live legacy model in the same process (the FM-00 (#1565) differential check). FM-2 (#1587) gates every published-artifact conversion on this hook against the Phase 0 goldens.

**Task:**
1. Generalize FM-00 (#1565)'s extractor from the tiny anchor to any supported legacy energy-family checkpoint: walk the pickled model, map every tensor to the canonical layout + its `CanonicalSpec`, record the full `extract_config_mace_model` field set + provenance, write safetensors + JSON. Fail loudly on any tensor it does not recognize — never silently drop weights — and refuse unsupported model classes by name per §4.
2. Add the end-user packaging mode: provision (or reuse) the pinned venv, run the extractor there as a subprocess, receive only the neutral file. **The extractor ships inside the `mace-torch` distribution as non-importable package data — a standalone script file, not a module** — executed by the pinned venv's own interpreter. End users therefore get it in the wheel, while no module under `packages/**` ever contains a static `import mace` and the two-entry double-import allowlist is unchanged. Development mode executes the same file against the in-tree frozen legacy package, so both modes run identical extraction code. Clear diagnostics when the pin cannot be installed.
3. Importer: neutral artifact → v1 model via the standard checkpoint-load path, applying the exact `full → reduced` projection (built from the symmetrized map) when the artifact declares `clebsch_gordan_basis: "full"`; reject `reduced → full` requests with a clear error.
4. Implement the golden verification hook (harness-JSON reference in; pass/fail with a per-quantity error report; dev-mode legacy self-check).
5. Expose one CLI entry point in `mace_torch.cli` covering both modes (working name `mace_convert_legacy`; final naming reviewable) — this is the command the migration guide (REL-1 (#1603)) will document.
6. Tests: round-trip on the P0-1 anchors (both modes), full-basis→reduced on a synthetic full-basis artifact, multi-head heads-intact, headless-source recording, unknown-tensor refusal, per-class refusal for the seven unsupported families, reduced→full refusal.

**Out of scope:** converting and publishing the actual foundation-model list (FM-2 (#1587)), MACE-Polar included; the model registry, naming/deprecation and loaders (FM-3 (#1588)); the artifact download/cache contract (FM-4 (#1636)); the migration guide text (REL-1 (#1603)); full-basis export for Symmetrix/libmace (DEP-3 (#1595)); deriving the reduced basis itself (CORE-4 (#1558)).

**Acceptance criteria:**
- [ ] Development mode converts the P0-1 tiny anchors in-process and the result reproduces the committed E/F/stress goldens at fp64 tolerance **and** matches the live legacy model loaded in the same process — FM-00 (#1565)'s gate, now passing through the production tool.
- [ ] End-user packaging mode, run on the same anchor checkpoint, produces canonical tensors bitwise-equal to the development-mode artifact (the pinned venv resolves to the same frozen legacy code).
- [ ] The sidecar's config field set matches `extract_config_mace_model` on the source model key-for-key (33 keys for the energy family, 32 for `AtomicDielectricMACE`); a missing field fails the conversion.
- [ ] A full-basis artifact is emitted/imported as reduced; the projection is built from the symmetrized map and reaches ≤1e-14 relative residual at fp64; the verification gate is on outputs, and no test asserts weight equality across the projection.
- [ ] v1's reduced basis has the pinned per-`(irrep_out, ν)` dimensions (`0e`: 1/4/8; `1o`: 1/3/12 for the published coupling at `correlation=3`); a dimension mismatch fails loudly rather than producing a lossy fit.
- [ ] `clebsch_gordan_basis` is read from the artifact (full- and reduced-tagged inputs take distinct, tested paths); `reduced → full` is refused with an actionable error; a requested basis whose machinery is unavailable is a hard error, never a silent substitution.
- [ ] A multi-head checkpoint converts with all heads intact; a headless checkpoint converts to `["default"]` with the headless origin recorded.
- [ ] Each of `AtomicDipolesMACE`, `EnergyDipolesMACE`, `PolarMACE`, `MACELES`, `MagneticMACE`, `MagneticScaleShiftMACE`, `MagneticSCFMACE` is refused by class with a message naming the owning ticket.
- [ ] An unrecognized tensor in the source checkpoint aborts the conversion with the tensor named — no silent drops.
- [ ] The neutral artifact contains no pickle; the step-2 importer imports no legacy `mace/` (import-direction check passes); metadata + provenance fields are present and round-trip.
- [ ] The golden verification hook fails the conversion on an injected tolerance breach.

**Verify:**
```bash
python -m pytest packages/mace-torch/tests/foundations/test_convert_legacy.py -v      # extractor/importer, basis rules, metadata, heads, refusals
python -m pytest packages/mace-torch/tests/foundations/test_cg_projection.py -v       # symmetrized-map projection, pinned reduced dimensions
python -m pytest tests/parity -k fm_convert -m "not slow"                             # dev-mode anchors vs goldens + live legacy in-process
python -m pytest packages/mace-torch/tests/foundations/test_convert_venv_mode.py -m network -v   # pinned-venv packaging path (installs mace-torch==0.3.17; opt-in)
```
**Review focus:** completeness of the weight walk and of the config field set (a silently dropped tensor or a dropped `avg_num_neighbors` is invisible until FM-2 (#1587)'s parity run), the projection being fitted on the symmetrized map rather than the raw reshape, the `full → reduced` gate being on goldens rather than weights, and the venv boundary — nothing pickled and no legacy import may reach the v1 side.

---
